domino/lib/src/experimental/idom_server.dart

```diff
@@ -0,0 +1,250 @@
+  void style(String name, String value) {
+    _shdElement.style[name] = value;
+  }
+
+  @override
+  void innerHtml(String value) {
+    _shdElement.nodes = [_IdomHtml(value, element)];
+    _index.last = 1;
+  }
+
+  @override
+  void skipNode() {
+    if (_index.last < element.nodes.length) {
+      _index.last = _index.last + 1;
+    }
+  }
+
+  @override
+  void skipRemainingNodes() {
+    _index.last = element.nodes.length;
+  }
+
+  @override
+  void event(String name, {fn, String key, bool tracked = true}) {
+    // TODO: implement event
+  }
+
+  void writeHTML(StringSink out,
+      {_IdomElem elem,
+      String indent,
+      bool indentAttr = false,
+      int indentLevel = 0}) {
+    elem ??= _rootElem;
+
+    // if elem.tag == null, then this elem is just a node list.
+    if (elem.tag == null) {
+      indentLevel = indentLevel - 1;
+    }
+    if (indent == null) indentAttr = false;
+    final indEnd = indent != null ? '\n' : '';
+    final curInd = (indent ?? '') * indentLevel;
+    final nextInd = (indent ?? '') * (indentLevel + 1);
+    final ml = indentAttr ? '\n$nextInd' : ' ';
+    final mml = indentAttr ? '\n$nextInd$indent' : ' ';
+
+    // tag generation, if indent is not null, ends on a new line
+    if (elem.tag != null) {
+      out.write('$curInd<${elem.tag}');
+      var simple = true;
+      if (elem.style.isNotEmpty) {
+        out.write('${ml}style="$mml');
+        out.write(elem.style.entries
+            .map((stl) => '${stl.key}: ${stl.value};')
+            .join(mml));
+        out.write('"');
+        simple = false;
+      }
+      if (elem.clazz.isNotEmpty) {
+        out.write('${ml}class="$mml');
+        out.write(_attrEscaper.convert(elem.clazz?.join(mml)));
+        out.write('"');
+        simple = false;
+      }
+      if (elem.attr.isNotEmpty) {
+        out.write(ml);
+        out.write(elem.attr.entries
+            .map((atr) => '${atr.key}="${_attrEscaper.convert(atr.value)}"')
+            .join(ml));
+        simple = false;
+      }
+      if (!simple) {
+        out.write(ml);
+      }
+      out.write('>$indEnd');
+    }
+
+    // wrting nodes, each ends in a new line if indent is not null
+    for (final node in elem.nodes) {
+      if (node is _IdomElem) {
+        // recursive element
+        writeHTML(out,
+            elem: node,
+            indent: indent,
+            indentAttr: indentAttr,
+            indentLevel: indentLevel + 1);
+      } else if (node is _IdomText) {
+        // text
+        out.write('$nextInd${_textEscaper.convert(node.text)}$indEnd');
+      } else if (node is _IdomHtml) {
+        // inline html block
+        out.write(node.html);
+      }
+    }
+
+    // closing tag
+    if (elem.tag != null) {
+      out.write('$curInd</${elem.tag}>$indEnd');
+    }
+  }
+}
+
+// Base class possible node types
+abstract class _IdomNode {
+  _IdomElem get parent;
+}
+
+// Element node
+class _IdomElem implements _IdomNode {
+  String tag;
+  String key;
```

Comment:
`key` can be ignored (and removed) here, as it only makes sense for DOM-reuse in the browser context.